## MA3K7: Big Mini Project rubric
### Problem 3: A pub game
#### Student ID: 2103654

#### Interactive game - play below!

In [1]:
# create n pint glasses
def create_glasses(n):
    return [0] * n

# printing out arrangement
def display(glasses):
    print("Glasses:", glasses)

# winning condition
def win_if(glasses):
    return any(glass > 1 for glass in glasses)

# what happens when Ali goes
def ali(glasses):
    while True:
        print("Ali's turn:")
        display(glasses)
        volumes = input("Enter the volumes of water to pour into the glasses (e.g., for n=4, '0.1 0 0.2 0.2'): ")
        volumes_list = [float(vol) for vol in volumes.split()]
        
        # want to make sure Ali adds exactly 1/2 pint during his turn
        if sum(volumes_list) == 0.5:
            
            # check if number of volumes entered is valid
            if len(volumes_list) <= len(glasses):
                # need to make sure allindices are valid
                valid_indices = all(0 <= vol < len(glasses) for vol in volumes_list)
                if valid_indices:
                    # add volumes to glasses
                    for idx, vol in enumerate(volumes_list):
                        glasses[idx] += vol
                    break
                else:
                    print("Invalid choice. Please enter valid indices.")
            else:
                print("Too many indices entered. Please enter at most {} indices.".format(len(glasses)))
        else:
            print("Volume needs to sum up to 0.5 pints. Please try again.")


# what happens when Beth goes
def beth(glasses):
    while True:
        print("Beth's turn:")
        display(glasses)
        
        # asks what glass to empty
        choice = input("Enter the index of the first glass to empty (0 to n-1): ")
        
        # makes sure input is valid
        if choice.isdigit():
            choice = int(choice)
            if 0 <= choice < len(glasses):
                glasses[choice] = 0
                
                # remove glass to the right of the one picked (including end case)
                glasses[(choice + 1) % len(glasses)] = 0
                break
            else:
                print("Invalid choice of index. Try again.")
        else:
            print("Invalid choice of index. Try again.")


# function to play the game
def main():
    n = int(input("Enter the number of glasses: "))
    glasses = create_glasses(n)

    while True:
        # players alternate turns
        ali(glasses)
        if win_if(glasses):
            display(glasses)
            print("Ali wins!")
            break
            
        beth(glasses)

if __name__ == "__main__":
    main()

Enter the number of glasses:  6


Ali's turn:
Glasses: [0, 0, 0, 0, 0, 0]


Enter the volumes of water to pour into the glasses (e.g., for n=4, '0.1 0 0.2 0.2'):  0.17 0 0.17 0 0.16


Beth's turn:
Glasses: [0.17, 0.0, 0.17, 0.0, 0.16, 0]


Enter the index of the first glass to empty (0 to n-1):  0


Ali's turn:
Glasses: [0, 0, 0.17, 0.0, 0.16, 0]


Enter the volumes of water to pour into the glasses (e.g., for n=4, '0.1 0 0.2 0.2'):  0.27 0 0.11 0 0.12


Beth's turn:
Glasses: [0.27, 0.0, 0.28, 0.0, 0.28, 0]


Enter the index of the first glass to empty (0 to n-1):  2


Ali's turn:
Glasses: [0.27, 0.0, 0, 0, 0.28, 0]


Enter the volumes of water to pour into the glasses (e.g., for n=4, '0.1 0 0.2 0.2'):  0.25 0 0 0 0.25


Beth's turn:
Glasses: [0.52, 0.0, 0.0, 0.0, 0.53, 0]


Enter the index of the first glass to empty (0 to n-1):  0


Ali's turn:
Glasses: [0, 0, 0.0, 0.0, 0.53, 0]


Enter the volumes of water to pour into the glasses (e.g., for n=4, '0.1 0 0.2 0.2'):  0 0 0 0 0.5


Glasses: [0.0, 0.0, 0.0, 0.0, 1.03, 0]
Ali wins!


### Finding minimum volume of water Ali needs to take from the bar each turn

In [2]:
import numpy as np

# know the the minimum value mu is between 0.45 and 0.5, but start from 0.4 for better idea of what's going on
xvals = np.linspace(0.4, 0.5, 10)

# only care about three glasses as per Ali's winning strategy
glasses=[0,0,0]

# function to find volume of water at the end of the game
def func(water):
    glasses = np.ones(3) * water
    
    # B empties glass 1
    glasses[0] = 0
    
    # find what volume at A2,  
    vol = water * 5/9
    
    # have Ali split water in half between two filled glasses 
    # Beth empties one of these
    # and then in next turn ALi adds all water to final cup
    final = (vol + water/2) + water
    
    print('A takes', "{:.4g}".format(water), 'pints of water. End volume =', "{:.5f}".format(final))

# array of water volumes Ali takes from bar
xvals = np.linspace(0.4, 0.5, 11)

for i in range(len(xvals)):
    func(xvals[i])
    

print('\n Look for the minimum volume:') 
func(0.4864)
func(0.4865)
func(0.4866)

print('\n so the minimum volume required is 0.4865 to 4 d.p.') 

A takes 0.4 pints of water. End volume = 0.82222
A takes 0.41 pints of water. End volume = 0.84278
A takes 0.42 pints of water. End volume = 0.86333
A takes 0.43 pints of water. End volume = 0.88389
A takes 0.44 pints of water. End volume = 0.90444
A takes 0.45 pints of water. End volume = 0.92500
A takes 0.46 pints of water. End volume = 0.94556
A takes 0.47 pints of water. End volume = 0.96611
A takes 0.48 pints of water. End volume = 0.98667
A takes 0.49 pints of water. End volume = 1.00722
A takes 0.5 pints of water. End volume = 1.02778

 Look for the minimum volume:
A takes 0.4864 pints of water. End volume = 0.99982
A takes 0.4865 pints of water. End volume = 1.00003
A takes 0.4866 pints of water. End volume = 1.00023

 so the minimum volume required is 0.4865 to 4 d.p.


## Extending

Looking at if A splits water equally between some amount of odd-numbered glasses:

In [3]:
# function that calculates sum [version 1]
def sum(n):
    total = 0
    for i in range(1, n+1):
        total += 1 / (2 * i)
    return total

def test10(k):
    array = np.arange(k, k+11, 1)
    for i in range(11):
        result = sum(array[i])
        print("Sum of 1/(2k) from k=1 to", array[i], "is:", result)

test10(25)
print('\n Initially thought that we would want to fill 31 glasses (total requires 62) but this forgets each glass can only hold one pint. Back to the drawing board...')

Sum of 1/(2k) from k=1 to 25 is: 1.9079790888767534
Sum of 1/(2k) from k=1 to 26 is: 1.9272098581075225
Sum of 1/(2k) from k=1 to 27 is: 1.9457283766260411
Sum of 1/(2k) from k=1 to 28 is: 1.963585519483184
Sum of 1/(2k) from k=1 to 29 is: 1.9808268987935287
Sum of 1/(2k) from k=1 to 30 is: 1.9974935654601953
Sum of 1/(2k) from k=1 to 31 is: 2.01362259771826
Sum of 1/(2k) from k=1 to 32 is: 2.02924759771826
Sum of 1/(2k) from k=1 to 33 is: 2.044399112869775
Sum of 1/(2k) from k=1 to 34 is: 2.0591049952227163
Sum of 1/(2k) from k=1 to 35 is: 2.0733907095084305

 Initially thought that we would want to fill 31 glasses (total requires 62) but this forgets each glass can only hold one pint. Back to the drawing board...


In [4]:
# adjusted function to calculate sum [version 2]
def newsum(n):
    total = 0
    for i in range(1, n-1):
        total += 1 / (2 * (i+2))
    return total

def test10(k):
    array = np.arange(k, k+11, 1)
    for i in range(11):
        result = newsum(array[i])
        print("Sum of 1/(2k) from k=1 to", array[i], "is:", result)

test10(3)

Sum of 1/(2k) from k=1 to 3 is: 0.16666666666666666
Sum of 1/(2k) from k=1 to 4 is: 0.29166666666666663
Sum of 1/(2k) from k=1 to 5 is: 0.3916666666666666
Sum of 1/(2k) from k=1 to 6 is: 0.4749999999999999
Sum of 1/(2k) from k=1 to 7 is: 0.5464285714285714
Sum of 1/(2k) from k=1 to 8 is: 0.6089285714285714
Sum of 1/(2k) from k=1 to 9 is: 0.664484126984127
Sum of 1/(2k) from k=1 to 10 is: 0.714484126984127
Sum of 1/(2k) from k=1 to 11 is: 0.7599386724386724
Sum of 1/(2k) from k=1 to 12 is: 0.801605339105339
Sum of 1/(2k) from k=1 to 13 is: 0.8400668775668775


AHA, we need A to fill 13 glasses during his first turn. After 11 rounds of playing and Beth emptying one filled glass each turn, there are just 3 glasses left filled. These will have 0.84006688 pints of water in each of them.

### A and C on the same team. Game play through.

In [9]:
c_n = 13

colin_glasses = np.ones(c_n) * 1/(2*c_n)

def update(array):
    print(array)
    k = len(array)
    ali = array + np.ones(len(array))* (1 / (2*(k-1)))
    beth = ali[1:]
    return beth

for i in range(1, c_n-1):
    print('Round', i)
    colin_glasses = update(colin_glasses)

Round 1
[0.03846154 0.03846154 0.03846154 0.03846154 0.03846154 0.03846154
 0.03846154 0.03846154 0.03846154 0.03846154 0.03846154 0.03846154
 0.03846154]
Round 2
[0.08012821 0.08012821 0.08012821 0.08012821 0.08012821 0.08012821
 0.08012821 0.08012821 0.08012821 0.08012821 0.08012821 0.08012821]
Round 3
[0.12558275 0.12558275 0.12558275 0.12558275 0.12558275 0.12558275
 0.12558275 0.12558275 0.12558275 0.12558275 0.12558275]
Round 4
[0.17558275 0.17558275 0.17558275 0.17558275 0.17558275 0.17558275
 0.17558275 0.17558275 0.17558275 0.17558275]
Round 5
[0.23113831 0.23113831 0.23113831 0.23113831 0.23113831 0.23113831
 0.23113831 0.23113831 0.23113831]
Round 6
[0.29363831 0.29363831 0.29363831 0.29363831 0.29363831 0.29363831
 0.29363831 0.29363831]
Round 7
[0.36506688 0.36506688 0.36506688 0.36506688 0.36506688 0.36506688
 0.36506688]
Round 8
[0.44840021 0.44840021 0.44840021 0.44840021 0.44840021 0.44840021]
Round 9
[0.54840021 0.54840021 0.54840021 0.54840021 0.54840021]
Round 10
[0

Total volume for 3 filled glasses =  0.84006688 x 3 = 2.52020064 > 2.5 pints required for Colin to redistribute so him and Ali win!